Better scala.js integration

betehess opened the pull request:
Includes:
- facade types for N3.js and jsonld.js
- `Future`-base `RDFReader`s

Remarks:
- jsonld.js is there as an internal dependency because the raw file had to be modified (just a bit) and we cannot use the webjar version for now (this will be fixed on the jsonld.js side)
- the zcheck tests do not make use of the Future-based framework for now (see https://github.com/InTheNow/zcheck/issues/12). This is not blocking in my opinion but others may disagree.

Finally, rdfstore.js was removed.

As this is important stuff, I am expecting reviews from @bblfish, @InTheNow, and @antonkulaga.

betehess committed:
[N3.js] First cut at exposing N3.js API.

Can be "tested" with
  testOnly org.w3.banana.n3js.N3Test

betehess committed:
[N3.js] Async parsing w/ error handling (Future-s)

betehess committed:
[N3.js] Added support for prefixes.

betehess committed:
[N3.js] Use js.Dictionary instead of introspecting properties.

betehess committed:
[N3.js] Removed unwanted scalajs-dom stuff.

betehess committed:
[N3.js] N3.js as an external dependency (WebJar)

betehess committed:
[N3.js] Moved facade types under a sub js package.

betehess committed:
[N3.js] Added a Future-friendly parse-by-chunk function.

betehess committed:
[N3.js] Implemented Banana-ready Turtle parser.

betehess committed:
[N3.js] Checks parser.addChunk is synchronous.

betehess committed:
[jsonld.js] Facade types for jsonld.js.

Important: the Javascript dependency is checked in because it had to be
modified to make sure that it worked in the Node.js environment. The
jsonld.js team is working on fixing that issue.

betehess committed:
Merge branch 'series/0.8.x' into wip/n3-js

Conflicts:
	project/build.scala

betehess committed:
[build] FastOptStage as default, remove rdfstore.js

betehess committed:
[build] Update travis.ci config

betehess committed:
[isomorphism] Fix erasure warning

betehess committed:
[jena] More time in JenaFusekiSparqlTest

betehess committed:
[build] More travis.ci

betehess committed:
[jena] JenaSparqlHttpEngine doesn't need to run in a Future.

betehess commented:
Also re-enabled travis.ci for `series/*`, fixed the build, improved build time, and fixed some warnings.

betehess committed:
[jsonld.js] Cleaning test

InTheNow commented:
>the zcheck tests do not make use of the Future-based framework for now (see InTheNow/zcheck#12). This is not blocking in my opinion but others may disagree.

@betehess assigned to task, as per  gitter chat "if you can give me some guidance I can give it a try"

bblfish commented:
We should also make sure to keep track of @antoniogarrote's contribution, as it seems like rdfstore is getting attention now as witness the discussion in this bug report https://github.com/antoniogarrote/rdfstore-js/issues/86 . As soon as it stabilises we will then be able to re-integrate the contribution .
Also note that I did some extra work on making access to rdfstorew much more direct ( i.e.. by removing the wrappers in the Scala eXchange branch of banana rdf that I published here https://github.com/read-write-web/banana-rdf/tree/eXchange2014

bblfish commented on .travis.yml:
```diff
@@ -1,51 +1,43 @@
+script:
+  - sbt ++$TRAVIS_SCALA_VERSION -jvm-opts travis/jvmopts test
+  # Tricks to avoid unnecessary cache updates
+  - find $HOME/.sbt -name "*.lock" | xargs rm
```
cool! How does this work? How does it avoid the updates? (I suppose this is regarding changes between VMs? )

betehess replied:
To be honest, I am not sure how it works. I just followed the doc at http://www.scala-sbt.org/0.13/docs/Travis-CI-with-sbt.html#%28Experimental%29+Reusing+Ivy+cache. See also the examples from http://docs.travis-ci.com/user/languages/scala/#Examples, especially the Spray one.

InTheNow replied:
IMHO, this is the opposite of what is required. A clean cache - ie no cache - will check that all required dependencies are brought in by _this_ build, rather than someone else's. eg: your build works, you push code, someone else pulls and the build fails. So, you delete your cache, rebuild and fails, so you add missing deps...

betehess replied:
Yes, there are pros and cons. If you look at the previous builds, it looks like this saves 5-10 minutes per build. The current build already takes well above 10 minutes.

betehess commented:
As per rdfstorew, I think it missed the opportunity to get ready to be included in the developing branch, and it was given plenty of time. It should now be considered as any other **new** feature: be ready in its own branch, and get added in a PR.

InTheNow commented on .travis.yml:
```diff
@@ -1,51 +1,43 @@
 # doc: http://about.travis-ci.org/docs/user/languages/scala/
```
Not sure I'm happy for this comment to be removed - I recall spending _alot_ of time getting sbt to work with scalajs and documented results. All the more so, given that the settings have been altered that go against the comments, and indeed one day ago there was an issue to build failures based on these comments.

InTheNow commented on N3.js/src/main/scala/org/w3/banana/n3js/io/N3jsTurtleParser.scala:
```diff
@@ -0,0 +1,45 @@
+package org.w3.banana
+package n3js
```
package names does not match naming convention - eg see https://github.com/w3c/banana-rdf/issues/139

betehess replied:
the full package is still `org.w3.banana.n3js.io` and it lives in the `org/w3/banana/n3js/io` directory. Am I missing something?

InTheNow replied:
Sorry, I expressed my point badly. I really meant to say that it differs to the other implementations, say in sesame. That said, I believe it is _better_ than what is there , so perhaps they could be updated too? But not a *must*

betehess replied:
You mean having the package split in two? It is mostly equivalent except for imports. I believe that the package name matches the directories in sesame/jena.

InTheNow commented:
General comments/thought on directories:
- N3.js and jsonld.js should go under io
- their js directory should be removed, files moved to parent

the travis directory is "overkill" for one file. 

Suggestion: ```git mv bin scripts``` and use scripts as a general directory for thing like this, example configs etc. Or perhaps use 'etc'

InTheNow commented on jsonld.js/src/main/scala/org/w3/banana/jsonldjs/js/JsonLdJs.scala:
```diff
@@ -0,0 +1,99 @@
+
+// the content of foo should be in jsonld but it doesn't work! Don't
+// know why...
+object foo {
```
have you tried with 0.6?

betehess replied:
I did in https://github.com/w3c/banana-rdf/commit/669dd3955979e4cd222326267006f3696c952b47 but couldn't make it work. So I just renamed foo in jsonldHelper for now.

betehess committed:
[travis] Removed cache, added links to documentation

betehess committed:
[N3.js] Moved facades under the top package

betehess committed:
[jsonld.js] Moved facades under top package

betehess committed:
[jsonld.js] s/foo/jsonldHelper/

I tried to move the function from foo back into jsonld but this
generates runtime exceptions.

betehess commented:
I think have addressed most of the issues. Just have a look at the commit messages :-)

bblfish commented on N3.js/src/main/scala/org/w3/banana/n3js/N3.scala:
```diff
@@ -0,0 +1,194 @@
+import scala.scalajs.js
+
+object N3 extends js.Object {
+
```
N3 is a bit of a misnomer since it does not parse N3: http://www.w3.org/2000/10/swap/Primer.html
Btw. as it can also parse quads, we need a quad type in RDF.

betehess replied:
I do not understand the comment.

The project is called is called N3.js, hence `N3` being used here. `N3.js` can actually parse N3. The current implementation focuses on Turtle and the types are not intended to lie here.

The fact that this PR doesn't expose **all** the features from N3.js and jsonld.js is a different matter. The quad things is a wanted feature but please stay focused on this PR for now. If that PR doesn't prevent one from doing what you are talking about in the future, then this is just good to go.

bblfish replied:
yes, the project is called N3.js but does it actually parse the N3 syntax defined here?
  http://www.w3.org/TeamSubmission/n3/
Ie does it do graphs, rules, and support the path notations '.' and '^' ?
We should ask @RubenVerborgh

betehess replied:
I don't know more that what the README says. It read trig and graphs are part of the callback. In any case, it is out of our jurisdiction and out of this PR for sure.

bblfish replied:
Well the name is misleading if it does not actually do N3. So perhaps @RubenVerborgh can clarify this.

RubenVerborgh replied:
It's called N3.js because this was my initial (and ultimate) implementation goal; however, as the [README](https://github.com/RubenVerborgh/N3.js/tree/v0.4.1#specifications) says:

> Note that the library does not support full Notation3 yet (and a standardized specification for this syntax is currently lacking).

I expect full N3 format somewhere this summer; although I should be careful with promises :smile:
The current parsed language is a superset of N-Quads and TriG—N3 will be supported some day.

bblfish replied:
ok then :-)

betehess replied:
It actually doesn't matter how @RubenVerborgh calls it. The name is the same as in his library, that's all.

RubenVerborgh replied:
_What's in a name? That which we call a rose
By any other name would smell as sweet._

bblfish replied:
> It actually doesn't matter how @RubenVerborgh https://github.com/RubenVerborgh calls it. The name is the same as in his library, that's all.
> 
> I’ll quote you on that :-)

betehess replied:
I do not understand your remark.

The only relevant information is the _Top-level JavaScript objects_ section in http://www.scala-js.org/doc/calling-javascript.html

bblfish replied:
Ah! I see. You can rename JS classes by using the `@JSName("Other")` annotation... but anyway, I was perhaps doing this code review a bit too quickly. This is not that important. I am happy that Ruben is going to implement N3 for real :-)

betehess replied:
I still don't get the comment. What **else** would you expect there?

bblfish replied:
Not sure really. N3 is a very specific syntax, that has a few parsers such as [cwm](http://www.w3.org/2000/10/swap/doc/cwm.html) and [euler](http://www.agfa.com/w3c/euler/). This seems to be a more general object that gives you access to a Parser, a Store and some Util libs that can process N3, Ntriples, Turtle, NQuads, ...

bblfish replied:
Note that N3 is really an implementation of Guha's thesis  "[Contexts 􏰄 A Formalization and Some Applications](http://www-formal.stanford.edu/buvac/guha-thesis.ps)"

betehess replied:
Whatever... Can we consider this "issue" closed for that PR?

bblfish replied:
"whatever" is a pretty dismissive remark, whereas this is perhaps the most important thing to understand about RDF. But ok, you can consider this ok, for this PR.

bblfish commented on jsonld.js/src/test/scala/org/w3/banana/jsonldjs/io/JsonLdJsParserTest.scala:
```diff
@@ -0,0 +1,39 @@
+import scalajs.concurrent.JSExecutionContext.Implicits.runNow
+import org.w3.banana.plantain._
+
+object JsonLdJsParserTest extends SpecLite {
```
Are we no longer doing generic tests? I thought we already had a test for JsonLD.

betehess replied:
Do we have a generic test? I mean with zcheck. As far as I know nobody has ported it. We have already discussed what that entails and that is a lot of work. For now, this is testing that the facades are correctly wired.

bblfish replied:
ok. Was just asking. :-)

InTheNow replied:
Why not just port over the JsonLd test (plus its dependencies)? At least then we are not adding an issue that this test will have to be deprecated/moved

betehess replied:
Do you mean the tests that rely on files on the filesystem?

InTheNow replied:
as far as i can see, JsonLDTestSuite.scala requires SerialisationTestSuite.scala. As far as i can see, neither rely on the FS

betehess replied:
I have just ported SerialisationTestSuite to zcheck before realising a writer instance was needed, which we don't have...

InTheNow replied:
Good that you have ported the test, one more of the list:) Re the missing writer.  Perhaps we should split SerialisationTestSuite into  reader/writer tests?

betehess replied:
At some point yes. But this goes beyond the scope of this PR.

InTheNow commented on .travis.yml:
```diff
@@ -1,51 +1,40 @@
+# * http://blog.travis-ci.com/2014-12-17-faster-builds-with-container-based-infrastructure/
 #
+# Java flags
+# * https://weblogs.java.net/blog/kcpeppe/archive/2013/12/11/case-study-jvm-hotspot-flag
```
typo- URL is missing final "s"

InTheNow replied:
Btw, the comments regarding java flags could/should be moved to .travis-jvmopts

betehess replied:
I don't know how to add comments in that file :-/

InTheNow replied:
usual "#*

betehess replied:
Fixed in the two travis.ci related files.

betehess committed:
[travis] Re-added comments

betehess committed:
[build] n3Js & jsonldJs are part of the main build

betehess committed:
[rdf-test-suite] Ported SerialisationTestSuite to zcheck

betehess committed:
[travis] 2Gb memory for Travis.

betehess committed:
[build] Move FastOptStage from global to jsonld-js

betehess committed:
[build] enable FastOptStage for plantain-js & n3-js

betehess committed:
[travis] Remove Xms settings

betehess committed:
Revert "[travis] Remove Xms settings"

This reverts commit 9061db1387c1c520acb22a4ba5f609536413a3b7.

betehess committed:
Revert "[build] enable FastOptStage for plantain-js & n3-js"

This reverts commit a33e5baef437fe906cf2aa88ff342332307d5dd9.

betehess commented:
@InTheNow @bblfish I believe I have now addressed all the comments. Can you please guy confirm and get in sync to finally merge this?

betehess committed:
[travis] Try without Xms again in that config

InTheNow commented:
:+1: from me

betehess commented:
Then the actual merge will be on @bblfish's plate :-)  I am assigning you, Henry!

sjrd commented on N3.js/src/main/scala/org/w3/banana/n3js/N3.scala:
```diff
@@ -0,0 +1,194 @@
+            promise.failure(ParsingError(error.get))
+          else {
+            prefixes.foreach { prefixes =>
+              js.Object.properties(prefixes).foreach((key: String) => prefixCallback(key, prefixes(key)))
```
This line could be instead:

``` scala
for ((key, prefix) <- prefixes)
  prefixCallback(key, prefix)
```

betehess replied:
Thanks!

betehess committed:
[n3-js] Use js.Dictionay#foreach
